src/server/service/diagnostics.rs

```diff
@@ -0,0 +1,690 @@
+            .and_then(|res| sink.success(res).map_err(Error::from))
+            .map_err(|e| debug!("Diagnostics rpc failed"; "err" => ?e));
+        ctx.spawn(future);
+    }
+
+    fn server_info(
+        &mut self,
+        _ctx: RpcContext<'_>,
+        _req: ServerInfoRequest,
+        _sink: UnarySink<ServerInfoResponse>,
+    ) {
+        unimplemented!()
+    }
+}
+
+mod log {
+    use std::convert::From;
+    use std::fs::{read_dir, File};
+    use std::io::{BufRead, BufReader, Seek, SeekFrom};
+    use std::path::Path;
+
+    use chrono::DateTime;
+    use nom::bytes::complete::{tag, take};
+    use nom::character::complete::{alpha1, space0, space1};
+    use nom::sequence::tuple;
+    use nom::*;
+    use rev_lines;
+
+    use futures::future::{err, ok};
+    use futures::Future;
+    use kvproto::diagnosticspb::{LogLevel, LogMessage, SearchLogRequest, SearchLogResponse};
+
+    const INVALID_TIMESTAMP: i64 = -1;
+    const TIMESTAMP_LENGTH: usize = 30;
+
+    struct LogIterator {
+        search_files: Vec<(i64, File)>,
+        currrent_lines: Option<std::io::Lines<BufReader<File>>>,
+
+        // filter conditions
+        begin_time: i64,
+        end_time: i64,
+        level: LogLevel,
+        filter: String,
+    }
+
+    #[derive(Debug)]
+    pub enum Error {
+        InvalidRequest(String),
+        ParseError(String),
+        SearchError(String),
+        IOError(std::io::Error),
+    }
+
+    impl From<std::io::Error> for Error {
+        fn from(err: std::io::Error) -> Self {
+            Error::IOError(err)
+        }
+    }
+
+    impl LogIterator {
+        fn new<P: AsRef<Path>>(
+            log_file: P,
+            begin_time: i64,
+            end_time: i64,
+            level: LogLevel,
+            filter: String,
+        ) -> Result<Self, Error> {
+            let log_path = log_file.as_ref();
+            let log_name = match log_path.file_name() {
+                Some(file_name) => match file_name.to_str() {
+                    Some(file_name) => file_name,
+                    None => {
+                        return Err(Error::SearchError(format!("Invalid utf8: {:?}", file_name)))
+                    }
+                },
+                None => {
+                    return Err(Error::SearchError(format!(
+                        "Illegal file name: {:?}",
+                        log_path
+                    )))
+                }
+            };
+            if log_name.is_empty() {
+                return Err(Error::InvalidRequest("Empty `log_file` path".to_owned()));
+            }
+            let log_dir = match log_path.parent() {
+                Some(dir) => dir,
+                None => {
+                    return Err(Error::SearchError(format!(
+                        "Illegal parent dir: {:?}",
+                        log_path
+                    )))
+                }
+            };
+
+            let mut search_files = vec![];
+            for entry in read_dir(log_dir)? {
```

Comment:
For reviewer: This PR is temporarily implemented using synchronous IO, which will later refactor asynchronous IO (implemented by me or by community members).  The reason is that we need to implement an asynchronous version of `rev_lines` dependency. Consider the modification of this PR is very huge, so split into two parts.